source/ada/lsp-lal_utils.adb

```diff
@@ -789,62 +811,104 @@ package body LSP.Lal_Utils is
             Text_Edit_Ordered_Maps.Next (Text_Edits_Cursor);
          end loop;
 
+         --  Resource operations are only supported if
+         --  `workspace.workspaceEdit.documentChanges` is True since they
+         --  must be sent in the `documentChanges` field.
+         --  `workspace.workspaceEdit.resourceOperations` client capability
+         --  must be checked in order to know which kind of operations are
+         --  supported.
+
          --  File creations
-         --  TODO
+
+         if Versioned_Documents and then Is_Create_Supported then
+            for File_Creation of Edits.File_Creations loop
+               WE.documentChanges.Append
+                 (LSP.Messages.Document_Change'(
+                  (Kind        => LSP.Messages.Create_File,
+                   Create_File => LSP.Messages.CreateFile'
+                     (uri    => LSP.Types.File_To_URI (File_Creation.Filepath),
+                      others => <>))));
+
+               declare
+                  Annotaded_Edits : LSP.Messages.AnnotatedTextEdit_Vector;
+                  Content : constant LSP.Messages.AnnotatedTextEdit :=
+                    LSP.Messages.AnnotatedTextEdit'
+                      (span    => ((0, 0), (0, 0)),
+                       newText => To_LSP_String
+                         (Ada.Strings.Unbounded.
+                            To_String (File_Creation.Content)),
+                       others  => <>);
+
+               begin
+                  Annotaded_Edits.Append (Content);
+
+                  WE.documentChanges.Append
+                    (LSP.Messages.Document_Change'(
+                     (Kind               => LSP.Messages.Text_Document_Edit,
+                      Text_Document_Edit => LSP.Messages.TextDocumentEdit'
+                        (edits => Annotaded_Edits,
+                         others => <>))));
+               end;
+            end loop;
+         end if;
 
          --  File deletions
 
-         if not Rename then
-            while Unbounded_String_Ordered_Sets.Has_Element
-              (File_Deletions_Cursor)
-            loop
+         if Versioned_Documents and then Is_Delete_Supported then
+            if not Rename then
```

Comment:
I usually swap `then` and `else` branches for `if not <Condition>` to simplify reading.
See an example in [Ada_Style_Guide](https://en.wikibooks.org/wiki/Ada_Style_Guide/Programming_Practices#Statements) under *Nesting* header.

Reply:
I agree.
In this case, the only difference between the two branches is:
`uri    => File_URI`
or
`uri    => File_URI & ".bak"`
Therefore, I'll remove the if statement, all the duplicated code, and do:
`(if Rename then File_URI & ".bak" else File_URI)`